Add NMR data model, integration with EnzymeML, and peak assignment features

torogi94 opened the pull request:
- Add [`md-models`](https://github.com/FAIRChemistry/md-models)-based research data model for NMR
  - Collect NMR parameters from instrument metadata
  - Record processing steps preformed in NMRpy
  - Save both the raw data and the latest version of the processed data
  - Assign species from a list or from an EnzymeML document to the peaks
- Add EnzymeML integration using [`pyenzyme`](https://github.com/EnzymeML/PyEnzyme) library
  - Load EnzymeML document into a FidArray
  - Have EnzymeML species available within Fid object
  - Save calculated concentrations to and return the EnzymeML document
- Add widgets to assign peaks either on the entire FidArray or single Fid objects
  - Use EnzymeML species if document available
  - Pass list of EnzymeML species if no document has been loaded yet
  - Use list of simple strings instead of working with EnzymeML standard
  - Use only a slice of Fids when assigning on the entire FidArray by passing an index list

torogi94 committed:
Initial commit

torogi94 committed:
Minor changes

torogi94 committed:
Implement IdentityAssigner for one FID object

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Implement IdentityAssigner for entire FIDArray

torogi94 committed:
Update data model

torogi94 committed:
Minor bug fix

torogi94 committed:
Add interface with EnzymeML

torogi94 committed:
Update IdentityAssigner

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Update generate_api.yaml

torogi94 committed:
Update generate_api.yaml

Commit:
API update

torogi94 committed:
Remove Citation info from data model

Commit:
API update

torogi94 committed:
Remove depricated data model

torogi94 committed:
Add apply_to_enzymeml() method

torogi94 committed:
Update plotting.py

torogi94 committed:
Restructure data model & reflect changes in lib

- Change FIDObject to be Peak-based instead of Identity-based
- Update PeakAssigner and PeakRangeAssigner to reflect new data model
- Add PeakIndexer and PeakRangeIndexer and add handlers to FID and FIDArray

torogi94 committed:
Delete .vscode directory

torogi94 committed:
Update from sdRDM to md-models

- Remove old sdRDM-based library
- Add new md-models-based library
- Consolidate PeakAssigner and IndentityAssigner into PeakAssigner
- Consolidate PeakRangeAssigner and IdentityRangeAssigner into PeakRangeAssigner
- Update Markdown specification for data model

torogi94 committed:
Temporarily remove conc. calculation

- Remove features from FidArray.calculate_concentrations() and raise NotImplementedError when called
- Remove features from ConcentraionCalculator and raise NotImplementedError when initialised

jmrohwer commented:
@torogi94 Thanks for this!
I need some time to look through this. But from a brief inspection the following come to mind:
- **Dependencies**: `utils.py` imports sympy as well as pyenzyme, yet none of them are declared in `requirements.txt`. We should discuss dependencies and which of these are needed for core functionality. One option is to have optional dependencies for certain functionalities. Pyenzyme itself has a multitude of dependencies. I am not keen for the core NMRPy library to have a huge list of dependences. Associated with that the version requirement dependency hell.
- Related: how stable are pyenzyme, md-models and the other dependencies from your group? E.g. Pyenzyme vs. Pyenzyme2? My concern is long-term maintenance.
- The conflict (above) needs to be resolved.
- Looking at `data_objects.py` as well as `plottling.py` there are a very large number of changes relating to linting and formatting (e.g. single vs. double quotes, linebreaks etc.). Are these linting changes done in a separate commit? As it is, it's very tedious to separate the *real* code changes from "cosmetic" linting.

torogi94 commented:
Thank you for looking over it! Here, a few comments:
> **Dependencies**: `utils.py` imports sympy as well as pyenzyme, yet none of them are declared in `requirements.txt`. We should discuss dependencies and which of these are needed for core functionality. One option is to have optional dependencies for certain functionalities. Pyenzyme itself has a multitude of dependencies. I am not keen for the core NMRPy library to have a huge list of dependences. Associated with that the version requirement dependency hell.

Good catch! I really did forget to declare the new dependencies in the requirements.txt file. I think you are right that we should not necessarily introduce them to the core of NMRpy. How about we declare them as optional nmrpy[enzymeml]? For the NMRpy data model, however, it would be easier if we could make pydantic a core dependency. If I am not mistaken, it is the only one required for the data model to work. If not, I can restructure my code, so the data model part becomes entirely optional, too, though.

> Related: how stable are pyenzyme, md-models and the other dependencies from your group? E.g. Pyenzyme vs. Pyenzyme2? My concern is long-term maintenance.

We are preparing a new stable pyenzyme (v2) release which we can then pin in the requirements. In my experience, it is already running better and will be more maintainable than the previous v1 version. With regard to md-models, the NMRpy data model is not dependent at all on the md-models version. Actually, it is not even a dependency needed, as all functionalities are handled via pydantic, which is itself a very well maintained library!

> The merge conflict needs to be resolved.

Absolutely. I’ll rebase my branch onto the latest main and fix the conflict manually so you can have a clean merge without conflicts!

> Looking at `data_objects.py` as well as `plottling.py` there are a very large number of changes relating to linting and formatting (e.g. single vs. double quotes, linebreaks etc.). Are these linting changes done in a separate commit? As it is, it's very tedious to separate the _real_ code changes from "cosmetic" linting.

Sorry for the noisy diff, it seems I accidentally let my linter loose on the two files and did not even notice... I’ll revert the linting changes from this PR and submit them in a separate PR (like discussed).

jmrohwer commented:
> Good catch! I really did forget to declare the new dependencies in the requirements.txt file. I think you are right that we should not necessarily introduce them to the core of NMRpy. How about we declare them as optional nmrpy[enzymeml]? For the NMRpy data model, however, it would be easier if we could make pydantic a core dependency. If I am not mistaken, it is the only one required for the data model to work. If not, I can restructure my code, so the data model part becomes entirely optional, too, though.

I agree with making pydantic a core dependency and the others as optional in `nmrpy[enzymeml]`.

> We are preparing a new stable pyenzyme (v2) release which we can then pin in the requirements. In my experience, it is already running better and will be more maintainable than the previous v1 version. With regard to md-models, the NMRpy data model is not dependent at all on the md-models version. Actually, it is not even a dependency needed, as all functionalities are handled via pydantic, which is itself a very well maintained library!

Okay great.

> Sorry for the noisy diff, it seems I accidentally let my linter loose on the two files and did not even notice... I’ll revert the linting changes from this PR and submit them in a separate PR (like discussed).

Give me a ping once you are ready with all of this then I'll proceed with the detailed code review :smile:

torogi94 committed:
Revert linting changes

torogi94 committed:
Update regular and add optional requirements

torogi94 committed:
Add graceful handling of optional imports

torogi94 committed:
Fix recurring typo in pyenzyme import error

torogi94 committed:
Remove deprecated data model linking

torogi94 committed:
Add None check for new FidArray properties

- Initialisation of enzymeml_document and concentrations properties of FidArray class with initial None value led to Error. Check for None added to fix this issue.
- Add correct optional dependency name to setup.py

torogi94 committed:
Fix various bugs

- Move the setup of Peak objects from the Peak(Range)Assigner to the deconvolution methods to prevent uninitialised Peak objects in data model.
- Update create_enzymeml() method to reflect changes in pyenzyme library.
- Add species property to FidArray, similar to the deconvoluted_integrals property.

torogi94 committed:
Update requirements.txt

Due to an issue with md-models, pydantic>=2.10.0 is currently causing errors. In the meantime, a version restraint has been added to the requirements.txt

torogi94 committed:
Update data_objects.py

Fix type checking bug in enzymeml_document property of FidArray class.

jmrohwer committed:
fix: Fix `Fid.baseline_correct()` and `FidArray.baseline_correct_fids()`

torogi94 committed:
Add measurement handling (#14)

* Add create_new_enzymeml_measurement() method

- FidArray now has a create_new_enzymeml_measurement() method that acts as the interface for creating new Measurement objects either per GUI or script.
- FidArray enzymeml_document property now checks for existence of at least one Measurement object in adder method.
- plotting.py now has a skeleton MeasurementCreator class that will later be the GUI for creating new Measurement objects.
- utils.py now has a utility function that creates an EnzymeML Measurement object from the parameters passed to it.

* Add multiple Measurement support

- Add MeasurementCreator widget as GUI for create_new_enzymeml_measurement() method
- Update apply_to_enzymeml() method
- Update create_enzymeml() method

torogi94 committed:
Rework data handling (#15)

* Update EnzymeML species handling

- Add species type flags to get_species_from_enzymeml() util function to allow filtering for specific types of species.
- Remove unnecessary display of proteins in PeakAssigner and PeakRangeAssigner using the new species type flags.

* Add flags for keeping data models upon saving

* Change data array handling of NMRpy data model

- Add save_data_model() method to serialise the NMRpy data model.
- Change handling of data arrays: They are now saved as numpy.ndarrays in each Fid object and only copied as lists into the data model upon serialisation.

* Update data_objects.py

- Resolve Pydantic serialisation issues (complex → string conversion).
- Optimise processing loops for faster execution and lower memory usage.

* Change keep_data_model flag to False

torogi94 commented:
I have addressed a few more issues we discussed in pull requests #14 and #15 and merged them with this pull request:
- With #14, handling of multiple EnzymeML measurements is now possible.
- With #15, data handling within NMRpy has been optimized to keep memory requirements and computation times low.

jmrohwer committed:
update ruff.toml

jmrohwer committed:
fix data_model.setter ; formatting fixes

jmrohwer committed:
Fix `FidArray.save_to_file()` when deleting data model

jmrohwer committed:
rework calibrate() widget not to make use of asyncio.Future()

jmrohwer committed:
revert @data_model.setter changes as data_model is now deleted during save, not set to None

jmrohwer committed:
fix pyenzyme imports, update dependencies

torogi94 committed:
Upgrade to pyenzyme v2

torogi94 committed:
Add t0 handling (#18)

- Added T0Tab and T0Logic classes to utils.py
- Added T0Adder widget class to plotting.py
- Added add_t0_to_enzymeml() method that can be used interactively with a Jupyter widget using gui=True or script-like with gui=False

torogi94 committed:
Fix copy() error in plotting.py

torogi94 committed:
Add explicit super init to FidArray

torogi94 committed:
Update range handling in Fid (#20)

Changed behaviour of Fid._setup_peak_objects(). Multiple peaks per range are handled properly now by using Fid._grouped_peaklist instead of Fid.peaks. Also refactored the method to make it more robust and readable overall.

torogi94 committed:
Add attribute checks for data model existence

torogi94 committed:
Fix faulty argument to T0Adder widget

torogi94 committed:
Add data model test suites

torogi94 committed:
Add data model property unit tests + fixes

torogi94 committed:
Fix parameter mapping to data model

- Added NMR parameters to fid_object setter
- Fixed p0 and p1 phasing parameter assignment

torogi94 committed:
Add proper single NMR in array edge case handling

torogi94 committed:
Update references to deprecated PyEnzyme features

torogi94 committed:
Add unit tests for new features

- Added tests for all of utils.py
- Added tests for new widgets in plotting.py
- Removed unused imports from nmrpy_tests.py

torogi94 committed:
Replace double ticks with single ticks

torogi94 committed:
Add documentation for new features

- Added information on optional installs to installation.rst
- Added data model information to intro.rst
- Added utils.py information to new utility_objects.rst
- Updated index.rst
- Described new features in quickstart.rst
- Added screenshots to _static/

jmrohwer committed:
convert build to use pyproject.toml

jmrohwer committed:
bump version to 1.0.0.dev0

jmrohwer commented on nmrpy/plotting.py:
```diff
@@ -1,3 +1,6 @@
+from typing import Mapping, Optional
+
+from pyenzyme.versions.v2 import Protein
```
@torogi94 Please look into this and confirm that you agree. Only usage of `Protein` is in line 1758. How will this affect the functionality if pyenzyme is not installed? Does the PeakAssigner class depend on pyenzyme? I'm assuming so, see https://github.com/NMRPy/nmrpy/blob/1928d7c478ab3b35f71f61d53599f16971ef01ef/nmrpy/data_objects.py#L1467-L1475

Thus seems okay to move the `Protein` import into the existing try/except block. Done.

jmrohwer commented on nmrpy/data_objects.py:
```diff
@@ -287,6 +318,87 @@ def ranges(self, ranges):
+    @property
+    def fid_object(self):
+        try:
+            self.__fid_object.raw_data = [str(datum) for datum in self.raw_data]
+            self.__fid_object.processed_data = [float(datum) for datum in self.data]
+        except AttributeError:
```
Calling `float(datum)` on a complex128 will not raise a `TypeError` but just issue a `ComplexWarning` and discard the imaginary part. Are we going to just ignore this error message? The point is, do you want the `FIDObject` to handle complex data in the processed_data list? This is an object that got introduced with the data model.

jmrohwer commented on nmrpy/data_objects.py:
```diff
@@ -604,18 +744,18 @@ def _phase_correct(cls, list_params):
                     )
             mz = lmfit.minimize(Fid._phased_data_sum, p, args=([data]), method=method)
-            phased_data = Fid._ps(data, p0=mz.params['p0'].value, p1=mz.params['p1'].value)
+            phased_data, p0, p1 = Fid._ps(data, p0=mz.params['p0'].value, p1=mz.params['p1'].value)
```
Not changed here but fixed `_ps()` to return degrees, so these phase angles are in degrees.

jmrohwer commented on nmrpy/data_objects.py:
```diff
@@ -604,18 +744,18 @@ def _phase_correct(cls, list_params):
             if verbose:
                 print('Zero order: %d\tFirst order: %d\t (In degrees)'%(mz.params['p0'].value, mz.params['p1'].value))
-            return phased_data
+            return phased_data, p0, p1
```
See previous comment, updated `_ps()` to return degrees.

jmrohwer commented on nmrpy/data_objects.py:
```diff
@@ -637,7 +777,7 @@ def _ps(cls, data, p0=0.0, p1=0.0):
             size = len(data)
             ph = numpy.exp(1.0j*(p0+(p1*numpy.arange(size)/size)))
-            return ph*data
+            return ph*data, p0, p1
```
Fixed by defining separate variables `p0rad` and `p1rad` for the phase angles in radians.

jmrohwer commented on nmrpy/data_objects.py:
```diff
@@ -1553,7 +1939,13 @@ def phase_correct_fids(self, method='leastsq', mp=True, cpus=None, verbose=True)
+                fid.data = datum[0]
+                # Update data model
+                if getattr(fid, 'fid_object', None) is not None:
+                    fid.fid_object.processed_data = [str(data) for data in datum]
```
@torogi94 I changed this to:
`fid.fid_object.processed_data = [str(d) for d in datum[0]]`
Am I right that the data model requires all arrays to be lists of strings?

jmrohwer commented on nmrpy/plotting.py:
```diff
@@ -1691,8 +1709,1166 @@ def assign(self):
+        for peak in self.fid.fid_object.peaks:
+            if peak.peak_position not in peak_position:
+                continue
+            peak.species_id = species_id.split(" ")[0]
```
@torogi94 How to deal with this one?

jmrohwer commented on nmrpy/plotting.py:
```diff
@@ -1691,8 +1709,1166 @@ def assign(self):
+            if peak.peak_position not in peak_position:
+                continue
+            peak.species_id = species_id.split(" ")[0]
+            fid.species[peak.peak_index] = peak.species_id
```
See previous.

jmrohwer commented on nmrpy/utils.py:
```diff
@@ -0,0 +1,612 @@
+            "A measurement ID is required to create an EnzymeML document. Please provide a measurement ID using the `measurement_id` keyword argument."
+        )
+    global_time = ([float(x) for x in fid_array.t],)
+    measurement = next(
+        measurement for measurement in enzymeml_document.measurements
+            if measurement.id == measurement_id
+    )
```
@torogi94 Please deal with this one. Seems reasonable at first sight.

jmrohwer commented on nmrpy/utils.py:
```diff
@@ -0,0 +1,612 @@
+        if "data_type" in kwargs:
+            try:
+                _data_type = pyenzyme.DataTypes[kwargs["data_type"].upper()]
+            except ValueError:
```
@torogi94 ?

jmrohwer commented on setup.py:
```diff
@@ -43,10 +43,11 @@
 NMRPy is developed by Johann Eicher and Johann Rohwer from the Laboratory for
 Molecular Systems Biology, Dept. of Biochemistry, Stellenbosch University, 
-South Africa.
+South Africa, as well as Torsten Giess from the Insitute of Biochemistry and
```
No longer relevant, `setup.py` has been removed.

jmrohwer committed:
apply fixes from Copilot code review

jmrohwer committed:
update README.md with extended description and pyproject.toml with additional classifiers

jmrohwer committed:
remove License from classifiers

jmrohwer committed:
fix nmrglue dependency

jmrohwer committed:
fix nmrglue dependency

jmrohwer committed:
fix numpy 2.4 deprecation bug and various typos

jmrohwer committed:
fix more imports when pyenzyme is not installed

jmrohwer committed:
update Anaconda and CI build scripts

jmrohwer commented on nmrpy/plotting.py:
```diff
@@ -1696,8 +1714,1166 @@ def assign(self):
+        for peak in self.fid.fid_object.peaks:
+            if peak.peak_position not in peak_position:
+                continue
+            peak.species_id = species_id.split(' ')[0]
```
Previous copilot comment (I think it got lost due to a force push, but it is still available in the comments pane, just not in the diff):
>Splitting on spaces truncates valid user-provided species names (the PR description explicitly mentions supporting 'simple strings'). This will mis-assign species like 'glucose 6 phosphate'. Instead, keep a stable mapping from dropdown display labels to the underlying species identifier/string (e.g., store the raw species value separately and avoid parsing the label).

My take on this is that it should be okay as IDs (in contrast to names) should not contain spaces. But do you check for this in the spec?

jmrohwer commented on nmrpy/plotting.py:
```diff
@@ -1696,8 +1714,1166 @@ def assign(self):
+        for peak in fid.fid_object.peaks:
+            if peak.peak_position not in peak_position:
+                continue
+            peak.species_id = species_id.split(' ')[0]
```
Previous Copilot comment (got lost as above):
>Same issue as in PeakAssigner: parsing the selected species by splitting on a space will corrupt any species string containing spaces. Use a label→value mapping (or dropdown options as (label, value) pairs) so the assigned value is exact.

Same point as above (ID vs name).

jmrohwer commented on nmrpy/plotting.py:
```diff
@@ -9,11 +11,28 @@
+try:
+    import pyenzyme
+    from pyenzyme import EnzymeMLDocument, Measurement
+    from pyenzyme.versions.v2 import Protein
```
I moved the import of `Protein` into the try-except block due to Copilot comment:
> This unconditional pyenzyme import will raise at import-time when pyenzyme isn't installed, even though the rest of the file treats it as an optional dependency. Move Protein into the existing try: block (or avoid importing it directly and use getattr(pyenzyme, ...) / isinstance(..., pyenzyme.Protein) inside the guarded section).

Need to verify that `Protein` is not required when pyenzyme is not available.
